### PR TITLE
COMPAT: avoid calling getsizeof() on PyPy

### LINKED ISSUE #17228: COMPAT: avoid getsizeof on PyPy
Pandas uses ``sys.getsizeof()`` to implement ``memory_usage()``. PyPy does not implement this functionality by design, since memory size is not fixed, depending on the JIT and other optimizations the memory usage of an object can change.

I would like to suggest one possible solution: using the default in sys.getsizeof(obj, default). If this is unacceptable, I could suggest that using ``deep=True`` would raise a ``TypeErrpr`` instead, but that would be more complicated. Any opinions? (proof-of-concept pull request forthcoming)

### PR DESCRIPTION
**Pull request by mattip**
closes #17228 

WIP Relates to issue #17228, this is a simple possible solution to the issue. Not sure this approach is the best, perhaps ``deep=True`` should raise on PyPy?

### CODE REVIEW

**Comment by chris-b1** [comment]
My first instinct would be to `raise` on PyPy with `deep=True` rather than stubbing a value, but don't feel strongly about it.

File: pandas/core/base.py [hunk 1]
```diff
@@ -23,6 +23,8 @@
 _indexops_doc_kwargs = dict(klass='IndexOpsMixin', inplace='',
                             unique='IndexOpsMixin', duplicated='IndexOpsMixin')
 
+import platform
+IS_PYPY = platform.python_implementation() == 'PyPy'
```
Comment (chris-b1) [inline pandas/core/base.py]: This can go in `pandas.compat`

**Comment by jreback** [comment]
not a big fan of downstream having to deal with this. I suppose using the default argument is ok.

File: pandas/util/testing.py [hunk 2]
```diff
@@ -56,6 +56,8 @@
 K = 4
 _RAISE_NETWORK_ERROR_DEFAULT = False
 
+import platform
```
Comment (jreback) [inline pandas/util/testing.py]: don't define this, just import from ``pandas.compat`` where you need it

**Comment by jreback** [comment]
pls add a whatsnew note in bug fixes maybe conversion section. ping on green.

**Comment by mattip** [comment]
@jreback ping. I added ``if PYPY`` conditional testing rather than just skipping entirely on PyPy where appropriate, those tests pass on PyPy too (PyPy latest nightly + Numpy HEAD is down to 44 failures).

File: doc/source/whatsnew/v0.21.0.txt [hunk 3]
```diff
@@ -307,8 +307,10 @@ Bug Fixes
 Conversion
 ^^^^^^^^^^
 
-- Bug in assignment against datetime-like data with ``int`` may incorrectly converte to datetime-like (:issue:`14145`)
+- Bug in assignment against datetime-like data with ``int`` may incorrectly convert to datetime-like (:issue:`14145`)
 - Bug in assignment against ``int64`` data with ``np.ndarray`` with ``float64`` dtype may keep ``int64`` dtype (:issue:`14001`)
+- Add a ``default`` argument to ``sys.getsizeof`` used in ``memory_usage`` to support PyPy. Objects on PyPy do not have a
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: spelling, remove the word guess. remove reference to ``sys.getsizeof``, this is not user visible.

File: pandas/core/indexes/multi.py [hunk 4]
```diff
@@ -467,7 +467,8 @@ def _nbytes(self, deep=False):
         """
         level_nbytes = sum((i.memory_usage(deep=deep) for i in self.levels))
         label_nbytes = sum((i.nbytes for i in self.labels))
-        names_nbytes = sum((getsizeof(i) for i in self.names))
+        objsize = 24  # for implementations with no useful getsizeof (PyPy)
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: put comments on a separate line (and skip a line)

File: pandas/core/indexes/range.py [hunk 5]
```diff
@@ -194,8 +194,11 @@ def _format_data(self):
 
     @cache_readonly
     def nbytes(self):
-        """ return the number of bytes in the underlying data """
-        return sum([getsizeof(getattr(self, v)) for v in
+        """ return the number of bytes in the underlying data
```
Comment (jreback) [inline pandas/core/indexes/range.py]: formatting, line up on the triple quote

File: pandas/tests/frame/test_repr_info.py [hunk 6]
```diff
@@ -332,13 +332,25 @@ def test_info_memory_usage(self):
         res = buf.getvalue().splitlines()
         assert re.match(r"memory usage: [^+]+$", res[-1])
 
-        assert (df_with_object_index.memory_usage(
-            index=True, deep=True).sum() > df_with_object_index.memory_usage(
-            index=True).sum())
```
Comment (jreback) [inline pandas/tests/frame/test_repr_info.py]: split into 2 tests, 1 for PYPY 1 for existing (skip for PYPY on existing, skip if not PYPY on other test)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. some doc / testing comments.

**Comment by mattip** [comment]
@jreback updated pull request according to the review

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mattip

**Comment by mattip** [comment]
thanks @jreback, @chris-b1. It is a pleasure to interact with such a responsive community

**Pull request closed without merging**